content/docs/rfcs/13/README.md

```diff
@@ -59,139 +59,260 @@ The following are not considered as part of the adversarial model:
+
+A `StoreResponse` without a populated `pagination_cursor` indicates that
+there are no more matching entries in the store.
+
+The client MAY request the next page of entries from the store service node
+by populating a subsequent `StoreRequest` with the `pagination_cursor` received in the `StoreResponse`. All other fields and query criteria MUST be the same as in the preceding `StoreRequest`.
+
+A `StoreRequest` without a populated `pagination_cursor` indicates that
+the client wants to retrieve the "first page" of the stored entries matching the query.
+
+## Store Request
+
+A client node MUST send all historical message queries within a `StoreRequest` message.
+This request MUST contain a `request_id`.
+The `request_id` MUST be a uniquely generated string.
+
+If the store client requires the store service node to include Waku message values in the query response,
+it MUST set `return_values` to `true`.
+If the store client requires the store service node to return only message hash keys in the query response,
+it SHOULD set `return_values` to `false`.
+By default, therefore, the store service node assumes `return_values` to be `false`.
+
+A store client MAY include query filter criteria in the `StoreRequest`.
+There are two types of filter use cases:
+1. Content filtered queries and
+2. Message hash lookup queries
+
+### Content filtered queries
+
+A store client MAY request the store service node to filter historical entries by a content filter.
+Such a client MAY create a filter on content topic, on time range or on both.
+
+To filter on content topic, the client MUST populate _both_ the `pubsub_topic` _and_ `content_topics` field.
+The client MUST NOT populate either `pubsub_topic` or `content_topics` and leave the other unset.
+Both fields MUST either be set or unset.
+A mixed content topic filter with just one of either `pubsub_topic` or `content_topics` set, SHOULD be regarded as an invalid request.
+
+To filter on time range, the client MUST set `time_start`, `time_end` or both.
+An unset `time_start` SHOULD be interpreted as "from the oldest stored entry".
+An unset `time_end` SHOULD be interpreted as "up to the youngest stored entry".
+
+If any of the content filter fields are set,
+namely `pubsub_topic`, `content_topic`, `time_start`, or `time_end`,
+the client MUST NOT set the `message_hashes` field.
+
+### Message hash lookup queries
+
+A store client MAY request the store service node to filter historical entries by one or more matching message hash keys.
+This type of query acts as a "lookup" against a message hash key or set of keys already known to the client.
+
+In order to perform a lookup query, the store client MUST populate the `message_hashes` field with the list of message hash keys it wants to lookup in the store service node.
+
+If the `message_hashes` field is set,
+the client MUST NOT set any of the content filter fields,
+namely `pubsub_topic`, `content_topic`, `time_start`, or `time_end`.
+
+### Presence queries
```

Comment:
In what way would you say this adds complexity? In fact, this is simply a hash based query and I'm just explaining this as one use case of hash-based queries. Trust assumptions are the same as for any other store query scenario. As for its value - this allows any client to query a store node for presence of a message hash known to the client, which makes it easy for e.g. lightpush clients to verify that their pushed messages have indeed been pushed to the network. Of course, the client would have to "trust" the store node or do redundant random checks to various store nodes to minimise probability of sybil attacks, but that is the case currently for all other store scenarios.

Reply:
> since client can verify the returned message match with requested hash, such attack is not possible.

But equally a malicious node may ignore the request forcing the client to resend the message or get confirmation from other store nodes. I agree that the client can also do "full checks" by simply requesting the entire message contents. The idea however is to provide a very lightweight method for clients to build reliability based on the redundancy and trust assumptions that make sense for their use case.
In every scenario there are two options:
1. Have some outside information that the node you're using is trustworthy - this is currently the case for all store node implementations and clients like Status Communities plan to use this model for a very long time (Community Owners set up their own trusted Store service nodes).  A lightweight confirmation mechanism makes sense for such a use case and this use case is going to be with us for the long term. The protocol should provision for such use case without prescribing for the specific client application context.
2. Use redundancy and randomization to prevent Sybil attacks, similar to how gossipsub etc. works. This would be the case once Store service is fully decentralized, if clients that choose to use this. For regular redundant checks a lightweight mechanism is necessary (a client shouldn't be expected to retrieve it's own published message multiple times from multiple nodes). This doesn't preclude more thorough checks where full message is downloaded and checked against a hash - it just describes one way in which the protocol can be used to build a lightweight "do you know about these set of messages?" mechanism.

> this is the reason why I think sync protocol should not only used between store nodes

I agree. The idea is to build this into a truly decentralized reliability layer.